In [1]:
import numpy as np
import time as t

# Question 1(a)

In [2]:
def gradient_decent(gradient, init_, learn_rate, n_iter = 50, tol = 1e-06):
    x = init_
    for _ in range(n_iter):
        delta = -learn_rate*gradient(x)
        if np.all(np.abs(delta) <= tol):
            break
        x += delta
    return round(x*1000)/1000

Gradient of $(x^2 + 3x + 4)$  is  $(2x + 3)$  
Gradient of $(x^4 - 3x^2 + 2x)$  is  $(4x^3 - 6x + 2)$

In [3]:
min1 = gradient_decent(lambda v:2*v + 3, 1.0, 0.2)
print("Minima of x^2 + 3x + 4  : " + str(min1))
min2 = gradient_decent(lambda v:4*v*v*v - 6*v + 2, 2, 0.1)
print("Minima of x^4 - 3x^2 + 2x  : " + str(min2))

Minima of x^2 + 3x + 4  : -1.5
Minima of x^4 - 3x^2 + 2x  : -1.366


# Question 1(b)

In [4]:
def gradient(w, X, y):
    a, b = w
    y_current = a * X + b
    N = len(X)
    a = -(2/N) * sum(X * (y - y_current))
    b = -(2/N) * sum(y - y_current)
    return np.array([a,b])

gradient() function is taking three parameter :  
w -> Touple of (a, b)  
X -> Values of x  
y -> actual values of y

It is returning a array of (a, b) which are the gradients of y = ax + b

# Question 1(c)

In [5]:
#Creating Data
np.random.seed(0)
X = 2.5 * np.random.randn(10000) + 1.5
res = 1.5 * np.random.randn(10000)
y = 2 + 0.3*X + res

In [6]:
#Little modified gradient decent function for linear regression
def gradient_decent_lreg(gradient, init_, learn_rate, n_iter = 50, tol = 1e-06):
    x, X, y = init_
    for _ in range(n_iter):
        delta = -learn_rate*gradient(x, X, y)
        if np.all(np.abs(delta) <= tol):
            break
        x += delta
    return np.round(x*1000)/1000

In [7]:
a , b = 0, 0
init = [(a, b), X, y]
a,b = gradient_decent_lreg(gradient, init, 0.01, 1000)
print("a = " + str(a) + "    b = " + str(b))

a = 0.295    b = 2.023


# Question 1(d)

In [8]:
#To divide the data if random mini batches of size B
def create_mini_batches(X, y, B):
    mini_batches = []
    data = np.stack((X, y), axis = 1)
    np.random.shuffle(data)
    n = len(X) // B
    for i in range(n):
        batch = data[i*B : (i+1)*B]
        mini_batches.append((batch[:,0], batch[:,1]))
    if len(X) % B != 0:
        batch = data[(i+1)*B:]
        mini_batches.append((batch[:,0], batch[:,1]))
    return mini_batches

In [9]:
def minibatch_gradient_decent(gradient, init_, learn_rate, n_iter = 50, tol = 1e-06):
    x, X, y, B = init_
    batches = create_mini_batches(X, y, B)
    for _ in range(n_iter):
        bidx = np.random.randint(len(batches))
        X_new, y_new = batches[bidx]
        delta = -learn_rate*gradient(x, X_new, y_new)
        if np.any(np.abs(delta) <= tol):
            break
        x += delta
    return np.round(x*1000)/1000

In [10]:
w = [0, 0]
B = 64
init = [w, X, y, B]
a,b = minibatch_gradient_decent(gradient, init, 0.01, 1000)
print("a : " + str(a) + " b : " + str(b))

a : 0.304 b : 2.019


# Question 1(e)

Time performance of Stochastic Gradient Descent(SGD) vs Mini-batch Stochastic Gradient Descent(MSGD) v Gradient Descent(GD)

In [14]:
#SGD      -> Mini batch with batch size = 1
w = [0, 0]
B = 1
init = [w, X, y, B]
tic = t.process_time()
a,b = minibatch_gradient_decent(gradient, init, 0.01, 10000)
toc = t.process_time()
print("It took " + str(toc - tic) + " secs")
print("a : " + str(a) + " b : " + str(b))

It took 0.28125 secs
a : 0.029 b : 1.828


In [15]:
#MSGD
w = [0, 0]
B = 1024
init = [w, X, y, B]
tic = t.process_time()
a,b = minibatch_gradient_decent(gradient, init, 0.01, 10000)
toc = t.process_time()
print("It took " + str(toc - tic) + " secs")
print("a : " + str(a) + " b : " + str(b))

It took 0.109375 secs
a : 0.307 b : 1.981


In [16]:
#GD
w = [0, 0]
init = [w, X, y]
tic = t.process_time()
a,b = gradient_decent_lreg(gradient, init, 0.01, 10000)
toc = t.process_time()
print("It took " + str(toc - tic) + " secs")
print("a : " + str(a) + " b : " + str(b))

It took 1.28125 secs
a : 0.295 b : 2.023


So, we can see SGD and MSGD are taking almost same time and GD is taking more time for the given data in same no of iterations. So, each iteration in SGD and MSGD is much faster than GD. But GD is more precise, where are SGD is least precise.

For optimal minibatch:

In [17]:
B = 32
for _ in range(10):
    w = [0, 0]
    init = [w, X, y, B]
    tic = t.process_time()
    a,b = minibatch_gradient_decent(gradient, init, 0.01, 1000)
    toc = t.process_time()
    print("It took " + str(toc - tic) + " secs")
    print("a : " + str(a) + " b : " + str(b))
    b *= 2

It took 0.0625 secs
a : 0.295 b : 2.026
It took 0.03125 secs
a : 0.328 b : 2.101
It took 0.046875 secs
a : 0.284 b : 1.997
It took 0.03125 secs
a : 0.33 b : 1.97
It took 0.046875 secs
a : 0.275 b : 2.054
It took 0.03125 secs
a : 0.272 b : 2.016
It took 0.03125 secs
a : 0.282 b : 2.033
It took 0.0625 secs
a : 0.339 b : 1.983
It took 0.03125 secs
a : 0.299 b : 2.015
It took 0.03125 secs
a : 0.259 b : 2.016


As we see, increasing batch size increases accuracy, but time per iteration also increases.

# Question 2

$i) p(Cold,Fever) = p(Fever|Cold) * p(Cold) = 0.02 * 0.307 = 0.00614$

$ii) p(Cough|Cold) = \frac{p(Cough,Cold)}{p(Cough)} = \frac{p(Cough, Cold, Lung Disease) + p(Cough, Cold, !Lung Disease)}{p(Cough, Cold, Lung Disease) + p(Cough, Cold, !Lung Disease) + p(Cough, !Cold, Lung Disease) + p(Cough, !Cold, !Lung Disease)}$  
  
$ p(Lung Disease)= 0.1009 * 0.2 + 0.009*0.8 = 0.02098 $  
$ p(!Lung Disease)= 0.8991 * 0.2 + 0.999*0.8 = 0.97902 $  
$ p(Cough, Cold, Lung Disease) = 0.7525 * 0.02 * 0.02098 = 0.0003 $  
$ p(Cough, Cold, !Lung Disease) = 0.505 * 0.02 * 0.97902 = 0.0098 $  
$ p(Cough, !Cold, Lung Disease) = 0.505 * 0.98 * 0.02098 = 0.0103 $  
$ p(Cough, !Cold, !Lung Disease) = 0.01 * 0.98 * 0.97902 = 0.0095 $  
  
Using these values,  
$ p(Cough|Cold) = \frac{0.0003 + 0.0098}{0.0003 + 0.0098 + 0.0103 + 0.0095}$  
  
$p(Cough|Cold) = 0.3377 $

# Question 3

Multinomial distribution : Generalized Binomial distribution  
It models the probability of counts for rolling a K-sided die N times  
$\begin{align}
p(\mathbf{y}) &= {{n}\choose{x_1, ..., x_k}}\prod_{i=1}^k y_i^{x_i} \\
&= C \prod_{i=1}^k {y_i^{x_i}}
\end{align}$  
Log Likelihood:  
$\begin{align}
l(\mathbf{y}) = \log p(\mathbf{y}) 
&= \log \bigg( C \prod_{i=1}^k {y_i^{x_i}} \bigg)\\
&= \log C + \log \prod_{i=1}^k {y_i^{x_i}} \\
\end{align}$  
$\begin{align}
y_k = 1 - \sum_{i = 1}^{k-1} y_i
\end{align}$  
$\begin{align}
l(\mathbf{y}) = \log p(\mathbf{y}) 
&= \log C + \log \left (\prod_{i=1}^{k-1} {y_i^{x_i}} \right ) {y_k^{x_k}}\\
&= \log C + \log \left (\prod_{i=1}^{k-1} {y_i^{x_i}} \right ) \left ( {1 - \sum_{i=1}^{k-1} y_i} \right )^{x_k} \\
&= \log C + \sum_{i=1}^{k-1} {x_i \log y_i} +  x_k \log \left ( {1 - \sum_{i=1}^{k-1} y_i} \right ) \\
\end{align}$  
$\begin{align}
\mathbf{\frac{\partial l}{\partial x_j}} &= 0
\Rightarrow0 + \frac{x_j}{y_j} + \frac{x_k}{\left ( 1 - \sum_{i=1}^{k-1} y_i \right )}\left ( -1 \right ) = 0\\
&\Rightarrow \frac{x_j}{y_j} = \frac{x_k}{\left ( 1 - \sum_{i=1}^{k-1} y_i \right)} = \frac{x_k}{y_k}\\
&\Rightarrow y_j = y_k*\frac{x_j}{x_k}
\end{align}$  
Now, we know  
$\begin{align}
\mathbf{y_1 + y_2 + .... + y_k} = 1
&\Rightarrow\left [\frac{x_1}{x_k} + \frac{x_2}{x_k} + ... + \frac{x_k}{x_k}\right ]y_k = 1\\
&\Rightarrow y_k = \frac{x_k}{\sum_{i = 1}^k x_i}
\end{align}$  
  
$y_k$ is the MLE for k-sided multinomial distribution